### PR TITLE
Support DeepSeek v3.2

### PR DESCRIPTION
**Pull request by zianglih**
@humansand

## Summary

This PR adds DeepSeek V3.2 support for Miles training and rollout, including the 5-layer debug/e2e workload used to validate MXFP8 mixed-BF16 training with SGLang rollout.

## Changes

- Add DeepSeek V3.2 model launch/config entries and `scripts/run_deepseek_v32.py` for prepare/train workflows.
- Add MBridge support for `deepseek_v32`, including DSA indexer weight mappings and V3.2/GLM-5 RoPE format handling.
- Add runtime HF config compatibility registration for clean `deepseek_v32` checkpoints, so we no longer mutate checkpoints with local shim files or patched `config.json` fields.
- Treat missing `indexer_rope_interleave` as non-interleaved/`False`; GLM-5 keeps its explicit interleaved setting from config.
- Add `--freeze-indexer` and supporting argument/config plumbing.
- Add MXFP8 mixed-BF16 defaults for V3.2 KV projection layers and torch-dist conversion pipeline-model-parallel handling.
- Add a B200 e2e test for `Pinaster/DeepSeek-V3.2-5layer` covering BF16 prepare, MXFP8 conversion, Megatron conversion, SGLang rollout, mixed-BF16 MXFP8 training, and fault-tolerance recovery.

## Validation

- Local hygiene: `pre-commit run --all-files` passed.
- Clean-checkpoint B200 devbox run passed after deleting all generated model/checkpoint directories and rerunning:
  `PYTHONPATH=/root/miles python3 tests/ci/run_suite.py --hw cuda --suite stage-c-8-gpu-b200 --labels run-ci-megatron`
- Result: `Test Summary: 1/1 passed`, runtime `1488.07s`.
- Metrics from the run: max `train/train_rollout_logprob_abs_diff=0.024676986038684845`, max `train/kl_loss=0.0006870048237033188`, `train/ppo_kl=0.0`.
- Post-cleanup smoke check on the clean BF16 checkpoint: `AutoConfig` resolves `DeepseekV32Config deepseek_v32 False`, and `AutoTokenizer` loads `LlamaTokenizer`.

### CODE REVIEW

File: tools/convert_hf_to_torch_dist.py [hunk 1]
```diff
@@ -65,9 +64,9 @@ def get_args():
     world_size = int(os.environ.get("WORLD_SIZE", "1"))
     args.global_batch_size = int(os.environ.get("WORLD_SIZE", "1"))
 
-    assert world_size <= args.num_layers, (
-        f"World size {world_size} must be less than or equal to number of layers {args.num_layers}. "
-        "You are using too many GPUs for this conversion."
+    assert args.pipeline_model_parallel_size <= args.num_layers, (
+        f"Pipeline model parallel size {args.pipeline_model_parallel_size} must be less than or equal to "
+        f"number of layers {args.num_layers}."
```
Comment (zianglih) [inline tools/convert_hf_to_torch_dist.py:49]: This is to relax the assertion to unblock conversion.

**Comment by Zhichenzzz** [comment]
Thanks @zianglih! Could you share which docker image and sgLang commit you used for testing?

**Comment by zianglih** [comment]
@Zhichenzzz I just saw your comment. It is just the miles dev image and the sglang inside.

**Comment by zianglih** [comment]
5layer mxfp8 verified by:
```
python scripts/run_deepseek_v32.py full-train   --no-enable-eval   --use-single-node   --hardware B200   --model-org Pinaster   --model-name DeepSeek-V3.2-5layer   --megatron-model-type deepseek-v32-5layer   --num-gpus-per-node 8   --actor-num-gpus-per-node 4   --rollout-num-gpus 4   --rollout-mxfp8   --train-mxfp8   --extra-args " \
    --use-rollout-routing-replay \
    --use-miles-router \
    --first-last-layers-bf16 \
    --num-layers-at-start-in-bf16 0 \
    --num-layers-at-end-in-bf16 1 \
    --sglang-disable-shared-experts-fusion \
    --num-rollout 3 \
  "
```
image

Since sglang has weird quantization override behavior for bf16 deepseek models, it is recommended to set `--num-layers-at-end-in-bf16 61` as bf16 workaround.

Note, we need a tiny sglang fix for `--first-last-layers-bf16`:
- https://github.com/sgl-project/sglang/pull/26287

**Comment by zianglih** [comment]
Manual B200 validation passed for the updated DeepSeek V3.2 5-layer MXFP8/BF16 e2e workload (`--global-batch-size 32` plus `--freeze-indexer`).

Command used on devbox `miles-v32-ci-963`:

```bash
cd /root/miles
export PYTHONPATH=/root/miles
python3 tests/ci/run_suite.py --hw cuda --suite stage-c-8-gpu-b200 --labels run-ci-megatron
```

Result: `1/1 passed` for `tests/e2e/megatron/test_deepseek_v32_5layer_mxfp8.py`.

Runtime: `1337s` as reported by `run_suite.py`.

Run id: `miles_v32_ci963_gbs32_freeze_indexer_20260526_002916`.

Key observed metrics remained in range through all 24 train steps: max `train/train_rollout_logprob_abs_diff` was about `0.02469`, max `train/kl_loss` was about `0.000692`, and `train/ppo_kl` stayed `0.0`. The rollout version check stayed clean (`rollout/weight_version/mixed_version_ratio=0.0`), including after the CI fault-injection recovery path.

Pushed validation commit: `12a8e8bd6`.

**Comment by zianglih** [comment]
Checkpoint shim cleanup status:

I tested whether the DeepSeek V3.2 checkpoint patching can be removed entirely, using a fresh clean model cache and `--num-rollout 1`. Before the no-shim run I deleted:

- `/root/models/DeepSeek-V3.2-5layer`
- `/root/models/DeepSeek-V3.2-5layer-bf16`
- `/root/models/DeepSeek-V3.2-5layer-MXFP8`
- `/root/models/DeepSeek-V3.2-5layer_torch_dist`

Result: the patching logic is still required for a clean unpatched `Pinaster/DeepSeek-V3.2-5layer` checkpoint in the current container. With checkpoint patching removed from the e2e prepare path, HF download plus BF16/MXFP8 conversion completed, but `tools/convert_hf_to_torch_dist.py` failed before training during tokenizer/config setup:

- `ValueError: The checkpoint you are trying to load has model type deepseek_v32 but Transformers does not recognize this architecture.`
- fallback then hit `AttributeError: 'PreTrainedConfig' object has no attribute 'max_position_embeddings'`

Failed run: `miles_v32_ci963_no_test_shim_clean_numrollout1_20260526_050817`, runtime 152s, `Test Summary: 0/1 passed`.

I also tested the narrower cleanup of deduplicating the test-local shim implementation by importing `_patch_deepseek_v32_checkpoint` from `scripts/run_deepseek_v32.py`. That passed from a clean model cache:

- run: `miles_v32_ci963_shared_shim_clean_numrollout1_20260526_051311`
- runtime: 730.25s
- `Test Summary: 1/1 passed`
- max `train/train_rollout_logprob_abs_diff`: 0.02460
- max `train/kl_loss`: 0.000703
- `train/ppo_kl`: 0.0

That result only proves the duplicate test shim can be removed; it does not remove the underlying launch-script patching. Since the actual patching is still needed for clean checkpoints, I am not treating the dedupe-only change as the desired cleanup.

**Comment by zianglih** [comment]
DeepSeek V3.2 shim cleanup note:

I tried removing the checkpoint patching from the 5-layer MXFP8 e2e test and validated from a clean cache by deleting `/root/models/DeepSeek-V3.2-5layer*` before rerunning with `--num-rollout 1`. That no-patch run failed during `convert_hf_to_torch_dist.py` prepare, before training:

- run id: `miles_v32_ci963_no_test_shim_clean_numrollout1_20260526_050817`
- result: `Test Summary: 0/1 passed`
- failure: `AutoConfig` / tokenizer loading still cannot resolve `model_type=deepseek_v32` in the installed Transformers, then the generic `PreTrainedConfig` fallback hits `AttributeError: max_position_embeddings`

So the patching is still required for clean DeepSeek V3.2 checkpoints. In particular, the BF16 checkpoint must be patched before HF-to-torch-dist conversion, and the MXFP8 checkpoint still needs the same remote-code metadata for serving.

As a control, I also ran from a clean cache with the test importing the shared launcher helper instead of carrying duplicate shim code:

- run id: `miles_v32_ci963_shared_shim_clean_numrollout1_20260526_051311`
- result: passed, `Test Summary: 1/1 passed`, runtime `730.25s`
- max observed `train/train_rollout_logprob_abs_diff`: `0.02460`
- max observed `train/kl_loss`: `0.000703`
- `train/ppo_kl`: `0.0`

Conclusion: removing the patching logic is not safe yet. The viable cleanup was only deduplicating test-local shim code against the launch script helper, but that was reverted so the PR currently keeps the explicit test patching.

**Comment by zianglih** [comment]
Milestone update: removed the DeepSeek V3.2 checkpoint patching path and validated the clean-checkpoint flow.

What changed:
- Replaced the ad hoc `DEEPSEEK_V32_CONFIG_SHIM` / `DEEPSEEK_V32_MODELING_SHIM` file mutation with runtime HF config compatibility registration.
- The new helper is an explicit alias registry, currently mapping `deepseek_v32` to the known-compatible Transformers `DeepseekV3Config` base. This keeps the design extensible for future unsupported-but-compatible model types without falling back to generic `PreTrainedConfig` behavior.
- Kept the V3.2 semantic default for missing `indexer_rope_interleave` in one runtime helper instead of writing it into checkpoint JSON.
- Removed patching from both `scripts/run_deepseek_v32.py` and `tests/e2e/megatron/test_deepseek_v32_5layer_mxfp8.py`.

Why this differs from the GLM5 cleanup in #1189:
- #1189 works for GLM5 because the stale GLM5 rewrite path was replaced by native `glm_moe_dsa` handling.
- DeepSeek V3.2 was failing earlier at HF config/tokenizer loading and at the missing `indexer_rope_interleave` attribute, so the correct cleanup is runtime config compatibility plus an explicit DSA default, not checkpoint mutation.

Validation:
- Commit pushed: `16b29c101 Avoid DeepSeek V3.2 checkpoint shims`.
- Clean-start validation deleted `/root/models/DeepSeek-V3.2-5layer`, `/root/models/DeepSeek-V3.2-5layer-bf16`, `/root/models/DeepSeek-V3.2-5layer-MXFP8`, and `/root/models/DeepSeek-V3.2-5layer_torch_dist`, then reran:
  `PYTHONPATH=/root/miles python3 tests/ci/run_suite.py --hw cuda --suite stage-c-8-gpu-b200 --labels run-ci-megatron`
- Result: `Test Summary: 1/1 passed`, runtime `1488.07s`.
- Metrics stayed in the expected range: max `train/train_rollout_logprob_abs_diff=0.024676986038684845`, max `train/kl_loss=0.0006870048237033188`, and `train/ppo_kl=0.0`.
- After the final registry refactor, I also smoke-checked the exact pushed helper on the devbox against the clean BF16 checkpoint: `AutoConfig` resolves to `DeepseekV32Config`, `model_type=deepseek_v32`, and `get_indexer_rope_interleave(...)` returns `False`.
- Local hygiene: `python3 -m compileall ...`, `git diff --check`, and `pre-commit run --all-files` passed.

**Review by yueming-yuan — APPROVED** [review]
